operators/src/adapters/raster_overlap_adapter.rs

```diff
@@ -0,0 +1,528 @@
+use crate::engine::{QueryContext, QueryRectangle, RasterQueryProcessor};
+use crate::error;
+use crate::util::Result;
+use futures::{ready, stream::BoxStream, StreamExt};
+use futures::{stream::Fold, Stream};
+use futures::{stream::FusedStream, Future};
+use geoengine_datatypes::error::Error::{GridIndexOutOfBounds, InvalidGridIndex};
+use geoengine_datatypes::{
+    primitives::{Coordinate2D, TimeInstance},
+    raster::{
+        Blit, CoordinatePixelAccess, GridIdx2D, GridIndexAccessMut, Pixel, RasterTile2D,
+        TileInformation, TilingStrategy,
+    },
+};
+use pin_project::pin_project;
+use std::task::Poll;
+
+use std::pin::Pin;
+
+type RasterStream<'a, T> = BoxStream<'a, Result<RasterTile2D<T>>>;
+type AccuType<T, X> = Result<(RasterTile2D<T>, X)>;
+
+pub type RasterFold<'a, T, FoldFut, FoldF, X> =
+    Fold<RasterStream<'a, T>, FoldFut, AccuType<T, X>, FoldF>;
+
+#[pin_project(project = RasterOverlapAdapterProjection)]
+pub struct RasterOverlapAdapter<'a, T, Q, QC, FoldF, FoldFut, XF, X, QF>
+where
+    T: Pixel,
+    Q: RasterQueryProcessor<RasterType = T>,
+    QC: QueryContext,
+{
+    query_rect: QueryRectangle,
+    time_start: Option<TimeInstance>,
+    time_end: Option<TimeInstance>,
+    // TODO: calculate at start when tiling info is available before querying first tile
+    spatial_tiles: Vec<TileInformation>,
+    current_spatial_tile: usize,
+    source: &'a Q,
+    query_ctx: &'a QC,
+    #[pin]
+    running_future: Option<RasterFold<'a, T, FoldFut, FoldF, X>>,
+    ended: bool,
+    fold_fn: FoldF,
+    accu_creator: XF,
+    tile_query_rewrite: QF,
+}
```

Comment:
1. I guess we need a comment on what this adapter does.
2. There are lots of generics, maybe we have to create slightly better (longer) names for them.

Reply:
added comment and simplified generics